pkg/build/gobuild.go

```diff
@@ -720,30 +742,41 @@ func (g *gobuild) buildOne(ctx context.Context, refStr string, base v1.Image, pl
 
 	var layers []mutate.Addendum
 
-	// Create a layer from the kodata directory under this import path.
-	dataLayerBuf, err := g.tarKoData(ref, platform)
-	if err != nil {
-		return nil, err
-	}
-	dataLayerBytes := dataLayerBuf.Bytes()
-	dataLayer, err := tarball.LayerFromOpener(func() (io.ReadCloser, error) {
-		return ioutil.NopCloser(bytes.NewBuffer(dataLayerBytes)), nil
-	}, tarball.WithCompressedCaching)
-	if err != nil {
-		return nil, err
+	// don't add kodata layer into wasi image
+	if platform.String() != "wasm/wasi" {
+		// Create a layer from the kodata directory under this import path.
+		dataLayerBuf, err := g.tarKoData(ref, platform)
+		if err != nil {
+			return nil, err
+		}
+		dataLayerBytes := dataLayerBuf.Bytes()
+		dataLayer, err := tarball.LayerFromOpener(func() (io.ReadCloser, error) {
+			return ioutil.NopCloser(bytes.NewBuffer(dataLayerBytes)), nil
+		}, tarball.WithCompressedCaching)
+		if err != nil {
+			return nil, err
+		}
+		layers = append(layers, mutate.Addendum{
+			Layer: dataLayer,
+			History: v1.History{
+				Author:    "ko",
+				CreatedBy: "ko build " + ref.String(),
+				Created:   g.kodataCreationTime,
+				Comment:   "kodata contents, at $KO_DATA_PATH",
+			},
+		})
 	}
-	layers = append(layers, mutate.Addendum{
-		Layer: dataLayer,
-		History: v1.History{
-			Author:    "ko",
-			CreatedBy: "ko build " + ref.String(),
-			Created:   g.kodataCreationTime,
-			Comment:   "kodata contents, at $KO_DATA_PATH",
-		},
-	})
 
 	appDir := "/ko-app"
-	appPath := path.Join(appDir, appFilename(ref.Path()))
+	var filename string
+	if platform.String() == "wasm/wasi" {
+		// module.wasm.image/variant=compat-smart expects the entrypoint binary to be of form .wasm
+		filename = appFilename(fmt.Sprintf("%v.wasm", ref.Path()))
+	} else {
+		filename = appFilename(ref.Path())
+	}
```

Comment:
```suggestion
	filename := appFilename(ref.Path())
	if platform.String() == "wasm/wasi" {
		// module.wasm.image/variant=compat-smart expects the entrypoint binary to be of form .wasm
		filename += ".wasm"
	}
```

It may be a silly nit, but I tend to prefer this since it makes it clearer that the wasi case is unusual, and normally this will just be `ref.Path()`